### licensed data
Apply for access from NIH, get the signature of PI, and then you can obtain the data. 

How to apply for access: https://dbgap.ncbi.nlm.nih.gov/aa/wga.cgi?page=pi_requests

```
/Share2/home/lulab1/TCGA
```
#### 1. miRNA-seq (bam file, hg38)
miRNA-seq of these cancers: BRCA, COAD, ESCA, LIHC, LUAD, LUSC, PAAD, READ, STAD
```
/Share2/home/lulab1/TCGA/miRNA-Seq
```

#### 2. RNA-seq (bam file, hg38)
RNA-seq of these cancers: BRCA, COAD, ESCA, LIHC, LUAD, LUSC, PAAD, READ, STAD
```
/Share2/home/lulab1/TCGA/RNA-Seq
```

#### 3. TCGA-L3 (counts, FPKM, hg38)
counts and fpkm of these cancers: BRCA, COAD, ESCA, LIHC, LUAD, LUSC, READ, STAD
```
/Share2/home/lulab1/TCGA/TCGA_L3
```


### public data

```
/Share2/home/lulab1/CCLE
```
#### 1. CellLineAnnotation (tsv)
Cell Line Annotations
```
/Share2/home/lulab1/CCLE/CellLineAnnotation
```
        
#### 2. RNAseq_Raw (bam, tsv)
RNAseq profiling of 1019 cell lines
```
/Share2/home/lulab1/CCLE/RNAseq_Raw
```
                
#### 3. mRNA expression Array (gct,res)
Raw Affymetrix CEL files were converted to a single value for each probe set using Robust Multi-array Average (RMA) and normalized using quantile normalization
```
/Share2/home/lulab1/CCLE/mRNAexpressionGeneExpression
```

#### 4. DNA Copy Number (gct, res)
Raw Affymetrix CEL files were converted to a single value for each probe set representing a SNP allele or a copy number probe.
```
/Share2/home/lulab1/CCLE/DNACopyNumberAffySNP
```

#### 5. Oncomap mutations (maf)
The mutations were assessed in 33 genes (381 specific mutations) using Oncomap 3.0 core
```
/Share2/home/lulab1/CCLE/OncomapMutations
```
    
#### 6. Hybrid capture sequencing (maf)
List of mutations and indels in 1651 genes, determined by targeted massively parallel sequencing.
```
/Share2/home/lulab1/CCLE/HybridCaptureSequencing
```

#### 7. Pharmacological profiling (maf)
Pharmacologic profiles for 24 anticancer drugs across 504 cell lines.
```
/Share2/home/lulab1/CCLE/PharmacologicalProfiling
```

